### PR TITLE
Improve list item parsing and indented code blocks parsing

### PR DESCRIPTION
**Pull request by LutSa**
What's in this PR:
1. solve the issue of #59 by using the technique of Open/Closed Blocks
2. change the UnorderedList type to a block container like BlockQuote
3. associate indent information to unordered list item and allow some nest specs according to the indent (and solve #81)
4. solve some minimal failing test cases about list item parsing by the way
5. fix minor failing test cases about “Indented code blocks” by allowing blank line between chunks

Update: How I  parse Lists and List items
1. parse the string to RawBlock type UnorderedListBlock/OrderedListBlock
	1. [Changes in [RawBlock.elm](https://github.com/dillonkearns/elm-markdown/pull/88/files#diff-c64e203e18fae05b2ee217cbeb744d92d8a2eda3270ac39617aab03ac432ca0eR24-R42) ] Firstly, we need to change the RawBlock type UnorderedListBlock/OrderedListBlock to contain “List RawBlock” instead of “List UnparsedInlines” because list item is container block that can contain other blocks. We also need to contain more information in the type UnorderedListBlock/OrderedListBlock to support Open/Close Block method: tight/loose, intended, marker. Because we parse the string to RawBlock when the block is closed, we also need to contain unparsed string as an open block that can merge later unparsed string. 
	2. [Changes in   [UnorderedList.elm](https://github.com/dillonkearns/elm-markdown/pull/88/files#diff-05709390a78385547d8f1efa8438f1947abfc96d8ec61ba5774ee22db3f2006b) and  [OrderedList.elm](https://github.com/dillonkearns/elm-markdown/pull/88/files#diff-dda95df11f221e8002e69e665d2b8495619d2fd63b6d4a9358ddc96a6336e57e)] Then, we need to change the parsers that parse string to some information that the RawBlock type UnorderedListBlock/OrderedListBlock need. Generally, I deleted the part where subsequent items are parsed because we only parse a single list item here and merge this list item by Open/Close Block method later.  And more information (intended and marker) is gotten. And there are also some minor fixes in order to fit the spec. (I’m sorry that the code style in these two files are bad.)
	3. [Changes in [RawBlock Parser unorderedListBlock and orderedListBlock in Parser.elm](https://github.com/dillonkearns/elm-markdown/pull/88/files#diff-de85daeee0fdb2d74f34329de78564a3902699d29f674d352a38d02dbab16b0dL407-L432)] Finally, the functions that convert the needed informations to RawBlock type UnorderedListBlock/OrderedListBlock also need to change.
2. apply Open/Close Block method to the parsing of UnorderedListBlock/OrderedListBlock.
	1. I made some changes in completeOrMergeBlock and completeBlock. I just applied the similar way of Open/Close Block method as we always use to merge list type. Changes in [completeOrMergeBlocks in Parser.elm](https://github.com/dillonkearns/elm-markdown/pull/88/files#diff-de85daeee0fdb2d74f34329de78564a3902699d29f674d352a38d02dbab16b0dR742-R807) (when the open block is UnorderedListBlock);  Changes in [completeOrMergeBlocks in Parser.elm](https://github.com/dillonkearns/elm-markdown/pull/88/files#diff-de85daeee0fdb2d74f34329de78564a3902699d29f674d352a38d02dbab16b0dR811-R877) (when the open block is OrderedListBlock); Changes in [completeOrMergeBlocks in Parser.elm ](https://github.com/dillonkearns/elm-markdown/pull/88/files#diff-de85daeee0fdb2d74f34329de78564a3902699d29f674d352a38d02dbab16b0dR925-R967)(when the open block is Blankline::OrderedListBlock or UnorderedListBlock); Changes in [completeBlocks](https://github.com/dillonkearns/elm-markdown/pull/88/files#diff-de85daeee0fdb2d74f34329de78564a3902699d29f674d352a38d02dbab16b0dR1171-R1248)
	2. Also, I match the intended of the open block and unparsed string to support intended list item [here](https://github.com/dillonkearns/elm-markdown/pull/88/files#diff-de85daeee0fdb2d74f34329de78564a3902699d29f674d352a38d02dbab16b0dR997-R1151).
3.   [Changes in [Block.elm](https://github.com/dillonkearns/elm-markdown/pull/88/files#diff-8a6bfd02a20644066e8a13c21d3e48f9f062e1021a032d07c6e6cd3113404940)]  Here, I just did some adjustments according to the type change (Inline -> Block in [Block type UnorderedList/OrderedList](https://github.com/dillonkearns/elm-markdown/pull/88/files#diff-8a6bfd02a20644066e8a13c21d3e48f9f062e1021a032d07c6e6cd3113404940L80-R81); UnparsedInline -> RawBlock in RawBlock type UnorderedList/OrderedList)
4.  [Changes in [Render.elm](https://github.com/dillonkearns/elm-markdown/pull/88/files#diff-d3a9446c640904e36e0b9d91324ab53d987f774ced108b29ea87e914c8c5e33b)] Here, we only need to did some adjustment saccording to the type change  (Inline -> Block in [Block type UnorderedList/OrderedList](https://github.com/dillonkearns/elm-markdown/pull/88/files#diff-8a6bfd02a20644066e8a13c21d3e48f9f062e1021a032d07c6e6cd3113404940L80-R81). We also need to render according to tight/loose. Thanks to Jonas’ work, I completely use his code here.

### CODE REVIEW

**Comment by dillonkearns** [comment]
It's nice that the actual renderer functions that users write won't actual need to change.
I discussed this design a while ago with @klaftertief and we could never fully decide on where the loose/tight list item rendering responsibility should be. Loose/tight handling be done at one of two levels:

1) within the user's renderer functions, or
2) in the code that calls the user's renderer functions

The implementation you have here uses approach (2). The logic lives outside the user's renderer functions. It's nice to see the full implementation, it makes it easier for me to understand which approach I like more. And seeing your implementation, my initial feeling is that I like this approach. It would seem very low-level if the user had to handle those details of loose/tight, and I imagine that most users would just find it frustrating, and wouldn't really find it useful to be able to customize how the render based on loose/tight, but instead would just want that detail taken care of in the implementation.

It looks like remark has a similar design. You can see in this example that they parse lists as `spread: true` for loose lists and `spread: false` for tight lists: https://astexplorer.net/#/gist/0a92bbf654aca4fdfb3f139254cf0bad/fe0a57596d677919245eca2b032c2a5ca892ec85.

And since the user can map over items in the parsed Markdown Blocks, they also have the option to transform blocks and change whether a list is loose or tight, that gives at least some customizability there if it's needed.

**Comment by dillonkearns** [comment]
For future reference, I thought of a 3rd design idea for how to handle loose/tight lists.

3. By default, automatically use the `paragraph` renderer to wrap loose paragraphs (as the current implementation does). Or, if a `looseListWrapper` is provided (`Just` value instead of `Nothing`), then use that to customize how to wrap loose lists.

I'm not sure there would be any value to approach (3) compared to the current approach, (2). I'm just including it here for future reference. If anyone can think of a use case where approach (3) would be beneficial, I would love to hear about it, though!

**Comment by LutSa** [comment]
I’ve let all test cases passed. But since I’ve changed the parser in UnorderedList.elm and OrderedList.elm from the parser of the list to the parser of the list item, the test cases may not be so meaningful as before. Please tell me if I can improve the test cases somehow.

**Comment by dillonkearns** [comment]
Excellent! I think it would be nice to have those unit tests exercising the full list parsing, so since that logic has moved up a level I made a commit to run the full parser on those test cases to keep them meaningful: d5d1e1334cf5c4b9dc7fcf7acd5cfc5895b505e1.

I think that's a good balance that will keep giving us confidence about those corner cases with the new parsing, so I think the tests are in good shape now 👍

File: src/Markdown/Block.elm [hunk 1]
```diff
@@ -77,8 +77,8 @@ In the simplest case, you can pass this directly to a renderer:
 type Block
     = -- Container Blocks
       HtmlBlock (Html Block)
-    | UnorderedList (List (ListItem Inline))
-    | OrderedList Int (List (List Inline))
+    | UnorderedList Bool (List (ListItem Block))
```
[the remark this answers is not in the document]
Comment (dillonkearns) [reply src/Markdown/Block.elm]: I went ahead and changed out the Bool's in the Block type, and I think it's looking good! So this is resolved now ✅

**Review by dillonkearns — CHANGES_REQUESTED** [review]

File: src/Markdown/ListItem.elm [hunk 2]
```diff
@@ -9,6 +9,7 @@ import Parser.Advanced as Advanced exposing (..)
 type ListItem
     = TaskItem Completion String
     | PlainItem String
+    | EmptyItem
```
[the remark this answers is not in the document]
Comment (dillonkearns) [reply src/Markdown/ListItem.elm:12]: Actually, I see that I was mistaken about this being user-facing. This is an internal type, so the user-facing types `Markdown.Block` aren't effected by this. So this is good as is 👍

File: test-results/failing/New/adjacent_lists.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+Should give output:
+
+````````````html
+<ul><li>This should be</li><li>An unordered list</li></ul><ol><li>This should be</li><li>An unordered list</li></ol>
+````````````
+
+But instead was:
+
+````````````html
+<ul><li>This should be</li><li>An unordered list</li></ul><ol></ol><ol start="2"></ol>
```
[the remark this answers is not in the document]
Comment (dillonkearns) [reply test-results/failing/New/adjacent_lists.md]: I figured out what was going on! It was an issue with the test runner. It looks like it wasn't deleting these files, so they got committed but don't didn't deleted even though they are passing now. This is in good shape now 👍

**Review by dillonkearns — APPROVED** [review]

**Comment by dillonkearns** [comment]
It's all looking good to me now! Thanks so much for the hard work @LutSa! 🎉
